src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java

```diff
@@ -377,12 +386,41 @@ private static <T> T checkNull(T obj) {
                 System.out.println(p11Info);
             }
 
-            if ((slotID < 0) || showInfo) {
+            if ((slotID < 0) || showInfo || (tokenLabel != null)) {
                 long[] slots = p11.C_GetSlotList(false);
                 if (showInfo) {
-                    System.out.println("All slots: " + toString(slots));
-                    slots = p11.C_GetSlotList(true);
-                    System.out.println("Slots with tokens: " + toString(slots));
+                    if (isZ) {
+                        System.out.println("Slots[slotID, tokenName]:");
+                        for (int i = 0; i < slots.length; i++) {
+                            System.out.println("[" + i + ", " + new String(p11.C_GetTokenInfo(slots[i]).label).trim() + "]");
+                        }
+                    } else {
+                        System.out.println("All slots: " + toString(slots));
+                        slots = p11.C_GetSlotList(true);
+                        System.out.println("Slots with tokens: " + toString(slots));
+                    }
+                }
+                /* TokenLabel is only supported on Z architecture platforms. It is null otherwise. */
+                if (tokenLabel != null) {
+                    boolean found = false;
+                    for (int i = 0; i < slots.length; i++) {
+                        String label = null;
+                        try {
+                            label = new String(p11.C_GetTokenInfo(slots[i]).label).trim();
+                            if (tokenLabel.equalsIgnoreCase(label)) {
+                                slotID = -1;
+                                slotListIndex = i;
+                                found = true;
+                                break;
+                            }
+                        } catch (PKCS11Exception ex) {
+                            label = null;
+                        }
+                    }
+                    if (!found) {
+                        throw new IOException("Invalid Token Label : "
+                                + tokenLabel);
+                    }
```

Comment:
I'm not a big fan of redundant variables. The `for` loop above could begin like this:
```java
                    for (int i = 0;; i++) {
                        if (i >= slots.length) {
                            throw new IOException("Invalid Token Label : "
                                    + tokenLabel);
                        }
```
(but I won't insist).